src/services/document-service.js

```diff
@@ -461,6 +474,67 @@ class DocumentService {
     }
   }
 
+  async regenerateDocument(documentKey) {
+    let lock;
+
+    try {
+      lock = await this.documentLockStore.takeLock(documentKey);
+
+      await this.transactionRunner.run(async session => {
+        const existingDocumentRevisions = await this._getAllDocumentRevisionsByKey(documentKey, session);
+
+        const document = this._buildDocumentFromRevisions(existingDocumentRevisions);
+
+        logger.info(`Saving document '${documentKey}' with revision ${document.revision}`);
+        await this.documentStore.save(document, { session });
+      });
+    } finally {
+      if (lock) {
+        await this.documentLockStore.releaseLock(lock);
+      }
+    }
+  }
+
+  async createRegenerateDocumentsBatch(user) {
+    const existingActiveBatch = await this.batchStore.findOne({
+      batchType: BATCH_TYPE.regenerateDocuments,
+      completedOn: null
+    });
+
+    if (existingActiveBatch) {
+      return null;
+    }
+
+    const batch = {
+      _id: uniqueId.create(),
+      createdBy: user._id,
+      createdOn: new Date(),
+      completedOn: null,
+      batchType: BATCH_TYPE.regenerateDocuments,
+      batchParams: {},
+      errors: []
+    };
+
+    const allDocuments = await this.documentStore.find({});
```

Comment:
This might not regenerate docs for all revisions (if there were missing docs for example). A safer way is to get all document keys by a unique query over `key` in the revisions collection. We can project to just return the keys.

Reply:
I thought we wanted to regenerate existing documents, okay, I'll do that.

Reply:
Fixed, now it selects from documentRevisions with "distinct" key.